Migrate read arguments and downstream code to GATKPathSpecifier

cmnbroad opened the pull request:
WIP. Some VCF outs are also converted, and probably a few other things as well.

cmnbroad commented:
@lbergelson @droazen This is the  next phase in URI migration. I'm planning to make more changes, but we can checkpoint this at any time (though I'd remove the TODOs I sprinkled throughout as reminders to myself for things I need to follow up on).

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java:
```diff
@@ -152,12 +153,12 @@
                                                              final ReadsSparkSource readsSource) {
         if (path == null) return null;
```
Can these paths be input as Path specifiers instead of strings here?

cmnbroad replied:
Yes, but it pulls on a longer thread. The pathseq package needs its own PR.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqScoreSpark.java:
```diff
@@ -136,13 +137,15 @@
 
     private int recommendedNumReducers = 0;
 
+    //TODO: fix this
```
looks like the same thread as above, I assume the todos mean we should hit them later instead of in this PR?  Should we open an issue?

cmnbroad replied:
Replaced these with https://github.com/broadinstitute/gatk/issues/6610 to track remaining todos.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java:
```diff
@@ -122,6 +132,7 @@ public static boolean isRemoteStorageUrl(String path) {
      * @param path the path
      * @return an absolute file path if the original path was a relative file path, otherwise the original path
      */
+    //TODO: get rid of this..
```
Let's burn this whole class with fire.

cmnbroad replied:
Yes, but still more work to do to get there.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java:
```diff
@@ -57,10 +57,10 @@
-    public static boolean isCramFile(final File inputFile) {
-        return isCramFileName(inputFile.getName());
+    public static boolean isCramFile(final GATKPathSpecifier pathSpec) {
+        return pathSpec != null && FileExtensions.CRAM.equalsIgnoreCase("." + FilenameUtils.getExtension(pathSpec.getURI().getPath()));
```
this needs a convenience method

cmnbroad replied:
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/ReadInputArgumentCollection.java:
```diff
@@ -30,18 +31,27 @@
+
+    /**
+     * Get the list of BAM/SAM/CRAM files specified at the command line.
+     * Paths are the preferred format, as this can handle both local disk and NIO direct access to cloud storage.
```
Fixed.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/ReadInputArgumentCollection.java:
```diff
@@ -30,18 +31,27 @@
 
     /**
      * Get the list of BAM/SAM/CRAM files specified at the command line.
      * Paths are the preferred format, as this can handle both local disk and NIO direct access to cloud storage.
```
Fixed.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/RequiredReadInputArgumentCollection.java:
```diff
@@ -16,28 +17,11 @@
-    }
+    /**
+     * Get the list of BAM/SAM/CRAM files specified at the command line.
+     * Paths are the preferred format, as this can handle both local disk and NIO direct access to cloud storage.
```
Fixed.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java:
```diff
@@ -323,31 +323,31 @@ public SAMFileHeader getHeaderForReads() {
         JavaRDD<GATKRead> output;
         // TODO: This if statement is a temporary hack until #959 gets resolve
-        if (input.endsWith(".adam")) {
+        if (inputSpecifier.getURIString().endsWith(".adam")) {
```
Fixed.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java:
```diff
@@ -504,20 +504,25 @@ public boolean useVariantAnnotations() {
 
     /**
-     * Returns the name of the source of reads data. It can be a file name or URL.
+     * Returns the name of the source of reads data. It can be a file name or URL. Throws if the tool has more
```
Yes, but lets do that in a separate PR.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java:
```diff
@@ -504,20 +504,25 @@ public boolean useVariantAnnotations() {
      */
-    protected LinkedHashMap<String, SAMFileHeader> getReadSourceHeaderMap(){
-        return readInputs;
+    protected SAMFileHeader getHeaderForInputPath(final GATKPathSpecifier inputPathSpecifier){
```
All I did here was to change the signature of the method so that we don't have to hand out the map. Instead the caller just queries it. So I think this change is orthogonal to the number of inputs.

cmnbroad replied:
Changed the name to `getHeaderForReadsInput`.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/SplitReads.java:
```diff
@@ -156,9 +157,9 @@ private SAMFileGATKReadWriter prepareSAMFileWriter(
-        final String base = FilenameUtils.getBaseName(readArguments.getReadFiles().get(0).getName());
-        final String extension = "." + FilenameUtils.getExtension(readArguments.getReadFiles().get(0).getName());
-        final File outFile = new File(OUTPUT_DIRECTORY, base + keyName + extension);
+        final String base = FilenameUtils.getBaseName(readArguments.getReadPathSpecifiers().get(0).getURI().getSchemeSpecificPart());
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/SplitReads.java:
```diff
@@ -156,9 +157,9 @@ private SAMFileGATKReadWriter prepareSAMFileWriter(
-        final File outFile = new File(OUTPUT_DIRECTORY, base + keyName + extension);
+        final String base = FilenameUtils.getBaseName(readArguments.getReadPathSpecifiers().get(0).getURI().getSchemeSpecificPart());
+        final String extension = "." + FilenameUtils.getExtension(readArguments.getReadPathSpecifiers().get(0).getURI().getSchemeSpecificPart());
+        final GATKPathSpecifier outFile = new GATKPathSpecifier(new File(OUTPUT_DIRECTORY, base + keyName + extension).getAbsolutePath());
```
DOne.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/RevertSamSpark.java:
```diff
@@ -269,7 +269,7 @@ protected void runTool(JavaSparkContext ctx) {
         Map<String, Path> writerMap = getOutputMap(outputMap,
                                                   output,
-                                                  getDefaultExtension(readArguments.getReadFiles().get(0).toString(), outputByReadgroupFileFormat),
+                                                  getDefaultExtension(readArguments.getReadPathSpecifiers().get(0).getRawInputString(), outputByReadgroupFileFormat),
```
Done.

cmnbroad commented:
Back to you @lbergelson:
- I added some new methods to `PathURI`  and/or `GATKPathSpecifier`  (`getExtension`/`hasExtension`, `getBaseName`, `isSam`, `isBam`, `isCram`, `isHadoopURL`). See what you think about the getExtension/hasExtension semantics - there is inherent weirdness/ambiguity around treatment of names that have multiple "." segments, like `some.fasta.gz`. The methods I added are mutually consistent, but maybe there is a better way.
- I also added https://github.com/broadinstitute/gatk/issues/6610 to track todos.
- Unfortunately I had to rebase this and make some more changes in order to get the merge build to pass.
- Once this goes in, I'll do the rename, and then continue with additional PRs.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSUtils.java:
```diff
@@ -57,12 +58,13 @@ public static void logItemizedWarning(final Logger logger, final Collection<?> i
     /**
      * Same as GATKSparkTool's getRecommendedNumReducers(), but can specify input BAM path (for when --input is not used)
      */
+    //TODO: fix this
```
It starts a cascade so it will be in a different pr.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/PathURI.java:
```diff
@@ -73,6 +76,100 @@ default String getScheme() {
+     * @throws IllegalArgumentException if the last component is empty (ie, the component ends in "/"), or the last
+     * component exists but starts with "."
+     */
+    default String getBaseName() {
```
We might want to prefer returning empty string when the base name/ extension are empty.   (not when they're missing because the file ends in "/" though...).

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/GATKPathSpecifierUnitTest.java:
```diff
@@ -286,6 +287,173 @@ public void testStdOut() throws IOException {
+
+                // no extensions
+                {"/", ".ext", false },
+                {".", ".ext", false },
```
Could you add a test for the case ".".hasExtension(".") ?

cmnbroad replied:
Hm, it doesn't. "." gets turned into a URI with an absolute path that ends in "/./".

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/GATKPathSpecifierUnitTest.java:
```diff
@@ -286,6 +287,173 @@ public void testStdOut() throws IOException {
+    public void testIsHadoopURL(final String referenceSpec, final boolean expectedIsHadoop) {
+        Assert.assertEquals(new GATKPathSpecifier(referenceSpec).isHadoopURL(), expectedIsHadoop);
+    }
+
```
Should we simple tests for isSam, isBam, is Cram ?

cmnbroad replied:
We've come this far so, yes.

lbergelson commented:
@cmnbroad Sorry about this.  I didn't really think about how nasty these methods might be when I asked for them in this pr.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathURI.java:
```diff
@@ -73,6 +76,100 @@ default String getScheme() {
+     * @throws IllegalArgumentException if the hierarchical name ends with the default file system separator
+     * (i.e. "/") or ".", or if the last component does not contain a ".".
+     */
+    default String getExtension() {
```
I modified this to return "" if there is no extension.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathURI.java:
```diff
@@ -73,6 +76,100 @@ default String getScheme() {
+        Utils.validateArg(extension.charAt(0) == '.', "Target extension must include the leading '.'");
+
+        // We don't want to use {@code #getExtension} here, since it won't work correctly if we're comparing an
+        // extension that uses multiple . chars, such as .fasta.gz., and {@code #getExtension} will throw if there
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathURI.java:
```diff
@@ -73,6 +76,100 @@ default String getScheme() {
+                }
+            }
+        }
+        throw new IllegalArgumentException(String.format("Input path (%s) has no extension", this));
```
Removed the throw and now returns "".

cmnbroad commented on src/test/java/org/broadinstitute/hellbender/engine/GATKPathSpecifierUnitTest.java:
```diff
@@ -286,6 +287,173 @@ public void testStdOut() throws IOException {
+                {"/."},
+                {"/name/.fasta"},
+                {"localFile"},
+                {"gs://hellbender/test/resources"},
```
Makes sense. Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathURI.java:
```diff
@@ -73,6 +76,100 @@ default String getScheme() {
+    default String getExtension() {
+        final String hierarchicalPath = getURI().getPath();
+        final int indexOfLastComponent = hierarchicalPath.lastIndexOf(FileSystems.getDefault().getSeparator());
+        if (indexOfLastComponent != -1 && indexOfLastComponent < hierarchicalPath.length() - 1) {
```
I don't *think* it is. The underlying URI always has a scheme with an (possibly empty) auth, and if the scheme is `file`, the path component is always an absolute path from the root, even if the input string was a relative path. There are tests for such cases, i.e., `localFile.bam` and `/localFile.bam`, so I think its ok.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/SplitReads.java:
```diff
@@ -146,20 +143,15 @@ private SAMFileGATKReadWriter createUnknownOutOnDemand(String attributeValue) {
-        final GATKPathSpecifier outFile = new GATKPathSpecifier(new File(OUTPUT_DIRECTORY, base + keyName + extension).getAbsolutePath());
+    private SAMFileGATKReadWriter prepareSAMFileWriter(final String keyName) {
+        final GATKPathSpecifier pathSpec = readArguments.getReadPathSpecifiers().get(0);
+        final GATKPathSpecifier outFile = new GATKPathSpecifier(
```
Added to https://github.com/broadinstitute/gatk/issues/6610.

cmnbroad commented:
So, to summarize the final basename/extension rules:

`getExtension` will throw if the path has no file name component (ie, it ends in '/'), otherwise it will return the shortest possible extension (i.e., ".gz" for "my.fasta.gz"), or "" if the name has no extension
`hasExtension` will return true for ".gz" or ".fasta.gz"
`getBaseName` will throw if the path has no file name component (ie, it ends in '/'), otherwise it will return the base of the file name component up to the last (possibly non-existent) "."

cmnbroad commented:
Ah, now it has conflicts...rebasing...

cmnbroad commented:
Need to resolve conflicts again...

cmnbroad commented:
Yeah last only (I think we only discussed a greedy one, but I never implemented it). Anyway, the only failures left are the SortSamSpark, so I'm merging. Next up - the GATKPathSpecifier name change.
